api/app/signals/apps/api/v1/fields/category.py

```diff
@@ -3,105 +3,58 @@
 from rest_framework import serializers
 from rest_framework.reverse import reverse
 
-from signals.apps.api.generics.relations import (
-    ParameterisedHyperLinkedIdentityField,
-    ParameterisedHyperlinkedRelatedField
-)
+from signals.apps.api.v1.fields.decorators import enforce_request_version_v1
 from signals.apps.signals.models import Category
 
 
-class ParentCategoryHyperlinkedIdentityField(serializers.HyperlinkedIdentityField):
-    lookup_field = 'slug'
+def category_public_url(category, request, format=None):
+    if category.is_child():
+        viewname, kwargs = 'public-subcategory-detail', {'parent_lookup_parent__slug': category.parent.slug,
+                                                         'slug': category.slug}
+    else:
+        viewname, kwargs = 'public-maincategory-detail', {'slug': category.slug}
+    return reverse(viewname, kwargs=kwargs, request=request, format=format)
 
-    def to_representation(self, value):
-        request = self.context.get('request')
-        result = OrderedDict([
-            ('curies', dict(name='sia', href=self.reverse('signal-namespace', request=request))),
-            ('self', dict(href=self.get_url(value, 'category-detail', request, None))),
-        ])
-
-        return result
-
-
-class CategoryHyperlinkedIdentityField(ParameterisedHyperLinkedIdentityField):
-    lookup_fields = (('parent.slug', 'slug'), ('slug', 'sub_slug'),)
 
+class CategoryHyperlinkedIdentityField(serializers.HyperlinkedIdentityField):
     def to_representation(self, value):
         request = self.context.get('request')
-        hyperlink = super(ParameterisedHyperlinkedRelatedField, self).to_representation(value=value)
         result = OrderedDict([
             ('curies', dict(name='sia', href=self.reverse('signal-namespace', request=request))),
-            ('self', {'href': hyperlink})
+            ('self', {'href': category_public_url(value, request)}),
         ])
-
         return result
 
 
-class CategoryHyperlinkedRelatedField(ParameterisedHyperlinkedRelatedField):
-    lookup_fields = (('parent.slug', 'slug'), ('slug', 'sub_slug'),)
-
-    view_name = 'category-detail'
+class CategoryHyperlinkedRelatedField(serializers.HyperlinkedRelatedField):
+    view_name = 'public-subcategory-detail'
     queryset = Category.objects.all()
 
+    def get_object(self, view_name, view_args, view_kwargs):
+        queryset = self.get_queryset()
+        if 'parent_lookup_parent__slug' in view_kwargs:
+            queryset = queryset.filter(parent__slug=view_kwargs['parent_lookup_parent__slug'])
+        return queryset.get(slug=view_kwargs['slug'])
 
-class LegacyCategoryHyperlinkedRelatedField(serializers.HyperlinkedRelatedField):
-    view_name = 'category-detail'
-    queryset = Category.objects.all()
+    def get_url(self, obj, view_name, request, format):
+        category = self.get_queryset().get(pk=obj.pk)
```

Comment:
No it can also be a `PKOnlyObject` so to make sure we always have a `Category` we need to fetch this. I added a check to see if this needs to be done. Because if it is already a `Category` we avoid making additional queries.